### DISCUSSION TITLE
RP2040 Bouncy_Circles example - fps

### DISCUSSION THREAD

**Discussion by elral** [Bodmer/TFT_eSPI#2095] [Raspberry Pi Pico]
I am playing around with your excellent library and have setup a Raspberry Pico with an ILI9341 Display from Waveshare.

Once connected I tried two examples from your library, the Bouncing_Circles and the RotatingCube. For the RotatingCube I am getting 80fps, for the Bouncy_Circle 17.9fps. Within the example it is written that it usually runs at 45-48fps, and I am wondering what I am missing to achieve this. As I am using PlatformIO, I have setup the platformio.ini like below. Changing the SPI frequnecy has no effect and using RP2040_PIO_SPI results in a compile error `.pio\libdeps\rpipico\TFT_eSPI\TFT_eSPI.cpp:994:10: error: 'dir_mask' was not declared in this scope`.
Do you have tips for me?
Thanks and regards
Ralf

[env:rpipico]
platform = https://github.com/maxgerhardt/platform-raspberrypi.git
board = rpipico
framework = arduino
board_build.core = earlephilhower	
board_build.filesystem_size = 0M
lib_ldf_mode = chain+
upload_protocol = mbed
lib_deps =
  bodmer/TFT_eSPI
  bodmer/TJpg_Decoder
build_flags =
    ; Define the TFT driver, pins etc here:
    -DUSER_SETUP_LOADED=1
    -DILI9341_DRIVER=1
    ;-DGC9A01_DRIVER=1
    ;-DTFT_HEIGHT=240
    ;-DTFT_WIDTH=240
    -DTFT_CS=12
    -DTFT_DC=13
    -DTFT_RST=14
    -DTFT_BL=15
    -DTFT_SCLK=18
    -DTFT_MOSI=19
    -DLOAD_GLCD=1
    -DLOAD_FONT2=1
    -DLOAD_FONT4=1
    -DLOAD_FONT6=1
    -DLOAD_FONT7=1
    -DLOAD_FONT8=1
    -DLOAD_GFXFF=1
    -DSMOOTH_FONT=1
    ;-DSPI_FREQUENCY=32000000
    -DSPI_FREQUENCY=80000000
    ;-DRP2040_PIO_SPI
monitor_speed = 115200

**Comment by Bodmer** [comment]
If the CPU clock is set higher than 125MHz then the maximum SPI clock rate is reduced to 24MHz by the board package (not TFT_eSPI,). This is done because the SPI peripheral is only specced to be clocked at 125MHz max.  (this means 62.5MHz maximum output rate because there is a divide by 2 in the SPI peripheral).

So to get the maximum SPI peripheral clock speed the processor must be run at 125MHz, this would then give ~45fps for the sketch.

Using PIO for SPI means the maximum. SPI clock rate is always half the CPU frequency even if it is overclocked.

It is possible to force the SPI peripheral to be overclocked using instructions in the sketch. I will post them if I can remember which sketch I tested them in.

Load the latest master copy of TFT_eSPI to get rid of the compile error.

**Comment by elral** [comment]
Thank's very much for your reply.
> So to get the maximum SPI peripheral clock speed the processor must be run at 125MHz, this would then give ~45fps for the sketch.

I guess it is a silly question, but how is it done? I looked aound in the Internet, but didn't find anything for the Pico.

Thanks and reagards Ralf

**Reply by Bodmer** [reply]
In the Arduino IDE there is a simple menu option to change the processor clock. I expect there is a software command but I have not investigated.

**Reply by Bodmer** [reply]
You can set the SPI peripheral clock to be equal to the CPU system clock by using the following lines:
```
 uint32_t  freq = clock_get_hz(clk_sys);
 // clk_peri does not have a divider, so in and out frequencies must be the same
 clock_configure(clk_peri,
                    0,
                    CLOCKS_CLK_PERI_CTRL_AUXSRC_VALUE_CLK_SYS,
                    freq,
                    freq);
```
This means that the maximum SPI output data clock will be half the CPU frequency (66.5MHz if the CPU clock is 133MHz).

**Comment by elral** [comment]
Thanks again for your reply.

I looked around and found how to set the system clock with earlephilhower's framework:
It has to be included:
`#include "pico/stdlib.h"`
and then setting the system clock can be done with:
`set_sys_clock_khz(125000, false);`
Within platformio.ini the SPI frequency and DMA use has to be defined:
`-DUSE_DMA_TO_TFT`
`-DSPI_FREQUENCY=62500000`
At least I had also to set the SPI frequency as you showed above.

Just using the PIO without setting the system clock to 125MHz haven't boost up as much as setting the system clock to 125MHz.

There is one drawback I realized when using PIO as SPI out and/or setting the system clock to 125MHz. A connectes servo does not work anymore. I haven't gone into detail what the reason is.

**Reply by Bodmer** [reply]
The SPI clock rate can only be set to an integer division of the processor clock. The library will drop the clock to the next lower nearest SPI frequency.

So, when the processor frequency is increased to say 133MHz then the maximum SPI rate is now 66500000. This means if you specify 62500000 as the frequency then the library will drop the rate to the next lowest value of 33250000, so you will see a speed drop.

One option is to set a higher than wanted frequency of say 100MHz, then it will not drop the clock frequency so far. However the TFT will probably not work reliably above a certain frequency. They typically run at a higher rate than the data sheet figures but this is on a case by case basis.

There are two PIO blocks, the PIO SPI required a complete block becuase I have added setAddress and block fill fnctions to the SPI code, so if the servo uses the PIO then maybe there are insufficient resources for PIO controlled servos.

The servo code requires [8 PIO instructions](https://github.com/earlephilhower/arduino-pico/blob/master/libraries/Servo/src/servo.pio.h#L34), 32 instruction locations are available in each block and there are 4 state machines in each block... so I would expect 4 servos to be possible. However I am not sure wat else might be using the PIO blocks.

The PIO SPI implementation uses[ all 32 instruction locations](https://github.com/Bodmer/TFT_eSPI/blob/master/Processors/pio_SPI.pio.h#L64).

**Comment by elral** [comment]
Thanks again, it's running perfect and fast!
Some question arise I will ask in a new issue.

Tried to mark it as resolved, but I can't change the label
